cmd/podman/attach.go

```diff
@@ -75,7 +75,7 @@ func attachCmd(c *cli.Context) error {
 		inputStream = nil
 	}
 
-	if err := attachCtr(ctr, os.Stdout, os.Stderr, inputStream, c.String("detach-keys"), c.BoolT("sig-proxy")); err != nil {
+	if err := startAttachCtr(ctr, os.Stdout, os.Stderr, inputStream, c.String("detach-keys"), c.BoolT("sig-proxy")); err != nil {
```

Comment:
I don't think we want `podman attach` to start stopped containers... Can you add a bool argument controlling whether to start the container when attaching?

Reply:
Actually the check is already done right before the call. But if you prefer to have an explicit parameter I am totally fine with it.

```go
	if conState != libpod.ContainerStateRunning {
		return errors.Errorf("you can only attach to running containers")
	}

	inputStream := os.Stdin
	if c.Bool("no-stdin") {
		inputStream = nil
	}

	if err := startAttachCtr(ctr, os.Stdout, os.Stderr, inputStream, c.String("detach-keys"), c.BoolT("sig-proxy")); err != nil {
```

Reply:
I think being explicit and having the extra check in `startAttachContainer` is probably safer, there are potential races with the current approach.

Reply:
>     return errors.Errorf("you can only attach to running containers")

This is an unnecessary limitation.  Attaching to a container and then, as a separate step, starting the container is a good way to make sure the attacher doesn't miss any bytes written after starting.  So I'm in favor of @mheon's ["should we start?" argument][1] or dropping `startAttachCtr` entirely and just making it a sequential `attachCtr` followed by a `startCtr`.

[1]: https://github.com/projectatomic/libpod/pull/1025#discussion_r199149513

Reply:
@wking The code above is the handler of the `podman attach` command, so checking that the container is running is something needed.

I am also in favor of having some common code doing the attach, and make if follow by a start for the `run` or `start` commands . However this impacts other files in libpod (`container_api.go`, `container_attach.go`, `attach.go`) so better make this change gradually.

Reply:
@wking Can't really do that, you run into races on slower systems. We needed to create an entirely new API endpoint to handle the ordering correctly (`StartAndAttach`) - can't do the operations seperately and expect that things will work on all systems

Reply:
I think the short answer is that `podman attach` is meant only for running containers, and if you need to attach to a not-running container, `podman start --attach` is the best option as it's guaranteed to do the right thing (hits the `StartAndAttach` API)

Reply:
> Can't really do that, you run into races on slower systems.

Where can I read about these races?  On slow systems it might take longer to make the `attach` connection.  But that just means you wait longer before calling `start`.  There's no rush to call `start` though, so I don't see a race.  Maybe the `attach` API was missing a "connection completed" event to trigger off of?

Reply:
@wking Yeah, we don't have a "Connection Completed" event fired out of `Attach` right now. We formerly used that, but it introduced a lot of code complexity (and bugs, as it turned out to be easy to make mistakes with how you called it). `StartAndAttach` simplified the signalling logic by ensuring we had an API call that was guaranteed to do the right thing.

Reply:
@mheon Is there any difference between `podman start --attach` over a running container vs `podman attach`?

Reply:
Thanks. Are there specific use cases for `start` being used for containers already running? I see t this behavior is in line with docker, but no other reasons.

Reply:
I think docker compat is the biggest reason. I suppose it's also handy for scripting? You can run a `podman start` to guarantee a container is up, without having to check its state beforehand

Reply:
> Are there specific use cases for `start` being used for containers already running? 

The runtime-spec [requires `start` to error if the container isn't in the `created` state][1].  I guess libpod tries to eat those errors?  It's not clear to me how you'd restart a stopped container though withought a racy release of the container ID...

[1]: https://github.com/opencontainers/runtime-spec/blob/v1.0.1/runtime.md#start

Reply:
@wking Our `Start` API is somewhat of a multipurpose beast. If the container is in `ContainerStateStopped`, IE has already run and stopped, we delete it from the runtime and clean up any artifacts of the last run and reset state to `ContainerStateInitialized`, meaning the container has not yet been created in the runtime. Then, if the container is in `ContainerStateInitialized` (either transitioned there from `ContainerStateStopped` or was there to begin with), we do an `Init()` to create the container in the runtime, and transition to `ContainerStateCreated`. Finally, for `ContainerStateCreated` containers (again, either just transitioned or in that state to begin with), we fire a `start` command through the OCI runtime. Containers not in these three states (notably, running containers) throw errors, which we catch on the `podman` side for Docker compatability. All of this is done under the same lock, so there should be no raciness.